core/drivers/pm/sam/at91_pm.c

```diff
@@ -0,0 +1,445 @@
+	int node = 0;
+
+	if (!set) {
+		io_write32(soc_pm.pmc + AT91_PMC_FSMR, mode);
+		return TEE_SUCCESS;
+	}
+
+	at91_sama5d2_config_shdwc_ws(soc_pm.shdwc, &mode, &polarity);
+
+	val = io_read32(soc_pm.shdwc + AT91_SHDW_MR);
+
+	/* Loop through defined wakeup sources. */
+	for (src = 0; src < ARRAY_SIZE(sama5d2_ws_ids); src++) {
+		wsrc = &sama5d2_ws_ids[src];
+		wsi = wsrc->info;
+
+		node = fdt_node_offset_by_compatible(soc_pm.fdt, -1,
+						     wsrc->compatible);
+		while (node >= 0) {
+			if (dev_is_wakeup_source(soc_pm.fdt, node)) {
+				/* Check if enabled on SHDWC. */
+				if (wsi->shdwc_mr_bit &&
+				    !(val & wsi->shdwc_mr_bit))
+					goto next_node;
+
+				mode |= wsi->pmc_fsmr_bit;
+				if (wsi->set_polarity)
+					polarity |= wsi->pmc_fsmr_bit;
+			}
+next_node:
+			node = fdt_node_offset_by_compatible(soc_pm.fdt, node,
+							     wsrc->compatible);
+		}
+	}
+
+	if (mode)
+		at91_sama5d2_config_pmc_ws(soc_pm.pmc, mode, polarity);
+	else
+		EMSG("AT91: PM: no ULP1 wakeup sources found!");
+
+	return mode ? TEE_SUCCESS : TEE_ERROR_BAD_STATE;
+}
+
+/*
+ * Verify that all the clocks are correct before entering
+ * slow-clock mode.
+ */
+static bool at91_pm_verify_clocks(void)
+{
+	int i = 0;
+	uint32_t scsr = 0;
+
+	scsr = io_read32(soc_pm.pmc + AT91_PMC_SCSR);
+
+	/* USB must not be using PLLB */
+	if ((scsr & (AT91SAM926x_PMC_UHP | AT91SAM926x_PMC_UDP)) != 0) {
+		EMSG("AT91: PM - Suspend-to-RAM with USB still active");
+		return false;
+	}
+
+	/* PCK0..PCK3 must be disabled, or configured to use clk32k */
+	for (i = 0; i < 4; i++) {
+		uint32_t css;
+
+		if ((scsr & (AT91_PMC_PCK0 << i)) == 0)
+			continue;
+		css = io_read32(soc_pm.pmc + AT91_PMC_PCKR(i)) & AT91_PMC_CSS;
+		if (css != AT91_PMC_CSS_SLOW) {
+			EMSG("AT91: PM - Suspend-to-RAM with PCK%d src %d", i,
+			     css);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static TEE_Result at91_write_backup_data(void)
+{
+	uint32_t val = 0;
+
+	while (1) {
+		val = io_read32(soc_pm.secumod + AT91_SECUMOD_RAMRDY);
+		if (val & AT91_SECUMOD_RAMRDY_READY)
+			break;
+	}
+
+	io_write32((vaddr_t)&at91bootstrap_bu->suspended, 1);
+	io_write32((vaddr_t)&at91bootstrap_bu->canary, virt_to_phys(&canary));
+	io_write32((vaddr_t)&at91bootstrap_bu->resume,
+		   virt_to_phys((void *)(vaddr_t)at91_pm_cpu_resume));
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result at91_enter_backup(void)
+{
+	int ret = -1;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = at91_write_backup_data();
+	if (res)
+		return res;
+
+	pm_change_state(PM_OP_SUSPEND, 0);
+	ret = sm_pm_cpu_suspend((uint32_t)&soc_pm,
+				(void *)at91_suspend_sram_fn);
+	if (ret < 0) {
+		DMSG("Suspend failed");
+		res = TEE_ERROR_BAD_STATE;
+	} else {
+		res = TEE_SUCCESS;
+	}
+
+	pm_change_state(PM_OP_RESUME, 0);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	/* SRAM content is lost after resume */
+	at91_pm_copy_suspend_to_sram();
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result atmel_pm_suspend(uintptr_t entry, struct sm_nsec_ctx *nsec)
```

Comment:
This function is not called from anywhere.

Reply:
Yes, I planned to add the glue in a next serie to enable that with PSCI (like what I did for reset and shutdown)

Reply:
Perhaps we can wait with this function until then, because it's not clear how it's going to be called right now.

Reply:
Hum, I think I'll add the glue layer in this series, it'll be simpler and the additionnal code is not really large.

Reply:
Sounds good